src/main/java/org/broadinstitute/hellbender/utils/python/PythonScriptExecutor.java

```diff
@@ -0,0 +1,208 @@
+package org.broadinstitute.hellbender.utils.python;
+
+import org.apache.commons.io.FileUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.broadinstitute.hellbender.exceptions.GATKException;
+import org.broadinstitute.hellbender.utils.Utils;
+import org.broadinstitute.hellbender.utils.io.IOUtils;
+import org.broadinstitute.hellbender.utils.io.Resource;
+import org.broadinstitute.hellbender.utils.runtime.ScriptExecutor;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * Generic service for executing Python Scripts.
+ *
+ * <ul><li>
+ * All tools that use PythonScriptExecutor must have a Java-based front-end, with standard GATK (Barclay-based) arguments.
+ * <li>
+ * Minimize the amount of code written in Python -- as much of each tool's work as possible should be done in Java. In
+ * particular, reading/writing final inputs and outputs should happen in Java. This is important to ensure the goal of
+ * universal GCS support, consistent Google authentication handling, etc.
+ *<li>
+ * The names of any files that are created by Python code should be passed in to the python code from Java.
+ *<li>
+ *<li>
+ * All dependencies (Python and native) of Python libraries used should be clearly documented, and included in the default
+ * GATK docker image.
+ * </li></ul>
+ *
+ * NOTE: Serial executions on the same PythonScriptExecutor are each run in a new process. No python state is retained
+ * between command/script/module execution. Using -i doesn't buy you anything (for this version of the executor, at
+ * least) since the process is terminated after each command completes.
+ */
+public class PythonScriptExecutor extends ScriptExecutor<PythonScriptExecutorException> {
+    private static final Logger logger = LogManager.getLogger(PythonScriptExecutor.class);
+
+    private static String python = "python";    // default executor name
+    private static String python3 = "python3";
+    public static String pyExtension = ".py";
+
+    private final List<String> curatedCommandLineArgs = new ArrayList<>();
+
+    /**
+     * @param ensureExists throw if the python executor cannot be located
+     */
+    public PythonScriptExecutor(boolean ensureExists) {
```

Comment:
Why do we expose this parameter here but not for the RScriptExecutor?

Reply:
It seemed like many/most of the RScriptExecutor use cases were for optional processing like post-traversal plotting, for which failure isn't catastrophic, but the Python use cases seem to have a hard requirement on Python. This was to make that dependency explicit and easy to express. Having said that, its makes sense to keep them as symmetric as possible, so I'll add a similar constructor to the RScriptExecutor.